feat(inpage-navigation): add component - INNO-1208

planctus opened the pull request:
# PR description

Please drop a few lines about the PR: what it does, how to test it, etc.

## QA Checklist

In order to ensure a safe and quick review, please check that your PR follow those guidelines:

* [ ] `package.json` is up-to-date and `@ecl/[system]-base` is part of the dependencies
* [ ] I have checked the dependencies
* [ ] I have given the fractal status “ready” to my component
* [ ] I have declared `@define mycomponent` in the SCSS file
* [ ] I have specified `margin: 0;` on the CSS component
* [ ] I have provided tests
* [ ] I follow the naming guidelines
* [ ] the component supports composition
* [ ] there are no hardcoded strings (all content come from the context)
* [ ] I have filled the README.md file (at least a few lines)
* [ ] My component is listed in the root README
* [ ] My PR has the right label(s)

HAL-Patch-INNO commented:
Deploy preview for *europa-component-library* ready!

Built with commit de999b6e14e4227efb4bac6ca3235a875cb1141e

https://deploy-preview-993--europa-component-library.netlify.com

emeryro commented on src/systems/ec/implementations/react/packages/inpage-navigation/InpageNavigation.jsx:
```diff
@@ -0,0 +1,95 @@
+import VanillaInpageNavigation from '@ecl/ec-component-inpage-navigation/ec-component-inpage-navigation';
+
+export default class InpageNavigation extends React.Component {
+  constructor(props) {
```
We try to avoid putting the call to the javascript behavior in the React component. On latest components, we call it in the index.jsx file (storybook). The idea is to rely less on React, and more on vanilla js.
Please have a look at the expandable or message component as an example.
It's true that all components haven't been updated yet with this new structure, sorry for the confusion.

emeryro commented on src/website/src/pages/ec/components/inpage-navigation/docs/code.mdx:
```diff
@@ -0,0 +1,11 @@
+import Playground from '@ecl/website-components/Playground';
+import InpageNavigationExample from '@ecl/ec-react-component-inpage-navigation/examples/Default';
+
+<Playground system="ec" selectedKind="InpageNavigation" selectedStory="default">
```
Link to the playground is not correct (you should make sure that the `selectedkind` is the same as the story name)

emeryro commented on src/systems/ec/implementations/vanilla/packages/ec-component-inpage-navigation/ec-component-inpage-navigation.scss:
```diff
@@ -0,0 +1,155 @@
+    text-align: left;
+    width: 100%;
+
+    &:focus {
```
there is no visual indication of the focus (and no hover). As it is up to the designers to provide it, please add a comment on the specs.

yhuard replied:
the focus state is very important
I've tried to use the component with the keyboard and couldn't tell that I was focus the button...
we need specs for this!

emeryro commented on src/systems/ec/implementations/react/packages/inpage-navigation/InpageNavigation.jsx:
```diff
@@ -0,0 +1,95 @@
+              className="ecl-inpage-navigation__trigger-icon"
+              shape="ui--corner-arrow"
+              size="s"
+              transform="rotate-180"
```
If we want to be coherant, the arrow should reverse when the dropdown is expanded (same as on the expandable). 
It's up to the designers to take the decision...

emeryro commented on src/systems/ec/specs/components/inpage-navigation/demo/data.js:
```diff
@@ -0,0 +1,25 @@
+    {
+      href: '#inline-nav-1',
+      label: 'Heading 1',
+      variant: 'standalone',
```
As the variant doesn't change, we should put it directly in the component's code.

emeryro commented on src/systems/ec/implementations/react/packages/inpage-navigation/package.json:
```diff
@@ -0,0 +1,22 @@
+  "devDependencies": {
+    "@ecl/ec-component-link": "^2.0.0",
+    "@ecl/ec-resources-icons": "^2.0.0",
+    "@ecl/ec-specs-link": "^2.0.0"
```
should be `ec-specs-inpage-navigation`

emeryro commented on src/systems/ec/implementations/react/packages/inpage-navigation/stories/Index.jsx:
```diff
@@ -0,0 +1,8 @@
+
+storiesOf('Navigation/In page navigation', module).add(
+  'default',
+  InpageNavigationExample
```
if possible, it would be nice to add knobs to edit inpage navigation labels (to test display on multiple lines for instance)

emeryro commented on src/systems/ec/implementations/vanilla/packages/ec-component-inpage-navigation/ec-component-inpage-navigation.js:
```diff
@@ -0,0 +1,138 @@
+  }
+
+  // Destroy
+  destroy() {
```
please remove event listener too on destroy.

yhuard commented on src/systems/ec/implementations/react/packages/inpage-navigation/examples/Default.jsx:
```diff
@@ -0,0 +1,60 @@
+import demoContent from '@ecl/ec-specs-inpage-navigation/demo/data';
+import InpageNavigation from '../src/InpageNavigation';
+
+const loremIpsum = require('lorem-ipsum');
```
why do you switch from `import` to `require`? is there any particular reason?

planctus replied:
yeah, a blind copy paste from the npm page for this package... :)

yhuard commented on src/systems/ec/implementations/react/packages/inpage-navigation/package.json:
```diff
@@ -0,0 +1,22 @@
+  "version": "2.1.1",
+  "description": "ECL EC React Inpage navigation",
+  "dependencies": {
+    "@ecl/ec-react-component-icon": "^2.1.1"
```
`@ecl/ec-react-component-link` is missing

yhuard commented on src/systems/ec/implementations/react/packages/inpage-navigation/package.json:
```diff
@@ -0,0 +1,22 @@
+    "@ecl/ec-react-component-icon": "^2.1.1"
+  },
+  "devDependencies": {
+    "@ecl/ec-component-link": "^2.1.1",
```
do you really need `@ecl/ec-component-link`?

planctus replied:
i love your rhetorical questions... :D

yhuard commented on src/systems/ec/implementations/react/packages/inpage-navigation/src/InpageNavigation.jsx:
```diff
@@ -0,0 +1,70 @@
+const InpageNavigation = ({ title, links, className, ...props }) => {
+  const classNames = classnames(className, 'ecl-inpage-navigation');
+  return (
+    <nav {...props} className={classNames} data-ecl-inpage-navigation-sticky>
```
why is it `data-ecl-inpage-navigation-sticky` and not just `data-ecl-inpage-navigation`? :)

yhuard commented on src/systems/ec/implementations/react/packages/inpage-navigation/src/InpageNavigation.jsx:
```diff
@@ -0,0 +1,70 @@
+        </button>
+        <ul
+          className="ecl-inpage-navigation__list"
+          aria-hidden="true"
```
are you sure you want to use `aria-hidden` and not `hidden`?

yhuard commented on src/systems/ec/implementations/react/packages/inpage-navigation/src/InpageNavigation.jsx:
```diff
@@ -0,0 +1,70 @@
+              <Link
+                {...link}
+                variant="standalone"
+                className={classnames('ecl-inpage-navigation__link')}
```
it's missing `link.className`

yhuard commented on src/systems/ec/implementations/react/packages/inpage-navigation/stories/Index.jsx:
```diff
@@ -0,0 +1,146 @@
+import VanillaInpageNavigation from '@ecl/ec-component-inpage-navigation';
+import InpageNavigation from '../src/InpageNavigation';
+
+const loremIpsum = require('lorem-ipsum');
```
same remark as before (`import` vs `require`)

yhuard commented on src/systems/ec/implementations/vanilla/packages/ec-component-inpage-navigation/ec-component-inpage-navigation.js:
```diff
@@ -0,0 +1,221 @@
+/**
+ * @param {object} options Object containing configuration overrides
+ */
+class InpageNavigation {
```
please `export class InpageNavigation` otherwise it won't be part of the vanilla JS build

planctus replied:
> the focus state is very important
I've tried to use the component with the keyboard and couldn't tell that I was focus the button

@yhuard i took a look at the demo in V1 and i couldn't notice differences regarding this, in the specs we have a generic reference "no changes" in the interaction, so i would assume we have to reproduce what we have in v1, most likely.

yhuard commented on src/website/src/pages/ec/components/inpage-navigation/index.md:
```diff
@@ -0,0 +1,5 @@
+---
```
I was expecting to find the page under "Navigation" on the website (there's already a blank page here)

yhuard commented on src/systems/ec/specs/components/inpage-navigation/package.json:
```diff
@@ -7,6 +7,9 @@
     "access": "public"
   },
+  "devDependencies": {
+    "lorem-ipsum": "^1.0.6"
```
please use an exact dependency (no caret)

yhuard commented on src/systems/ec/implementations/vanilla/packages/ec-component-inpage-navigation/ec-component-inpage-navigation.scss:
```diff
@@ -0,0 +1,159 @@
+    }
+  }
+
+  .ecl-inpage-navigation__link {
```
since you add this class to a `<Link>` component, you must make sure the styles from Link have already been imported. It's also true for the Icon component. It means:

1. add `@ecl/ec-component-link` and `@ecl/ec-component-icon` to the dependencies
2. add a check `@include check-imports(('ec-component-link', 'ec-component-icon'));` at the beginning of the file

yhuard commented:
one more thing:

![image](https://user-images.githubusercontent.com/3524389/53163709-f74c0400-35ce-11e9-9f88-488e005b1417.png)

the inpage nav should not appear before the first title (check v1)

planctus commented:
@yhuard , @emeryro , hope i didn't forget anything, many things had to be touched and i can't be entirely sure i covered everything you asked for.

- I updated the demo in the website adding site header, page header and footer.
- Fixed a bug in the js that was creating some visual bugs
- Added dependencies here and there, following your advices, then i added all the package i needed for the demo in the website as dev dependencies, hope it was the right thing
- Added the missing classnames for links
- Moved the website demo to the proper place, in the navigation block of components
- Changed the aria-hidden attribute in an html5 hidden one.
- Added missing checks in component's sass file imports
- Various other fixes

planctus commented:
I've added a template package, very basic one with only required elements, headings and dummy text with the inpage navigaiton in place, more or less the same we have already in the website demo, it is called "content page". ;)

emeryro commented:
A few points spotted ahead still need to be fixed:

- on mobile, when I open the inpage nav, leave it open and scroll to the top, the links stay on the page
- the first click on the inpage nav button doesn't do anything

planctus commented:
@emeryro, @yhuard, i fixed these two issues:

- first click on toggler in mobile 
- list appearing in mobile when the inpage list is open and the widget is removed by scrolling up (still the case in V1.. )

emeryro commented on src/systems/ec/implementations/react/packages/inpage-navigation/stories/Index.jsx:
```diff
@@ -0,0 +1,146 @@
+    button(btnIdLabel, btnIdHandler, groupId);
+    button(btnIdRemoveLabel, btnIdRemoveHandler, groupId);
+
+    const inpageProps = {
```
this should come from the demo.js file in the specs

emeryro commented on src/systems/ec/implementations/react/packages/inpage-navigation/examples/Default.jsx:
```diff
@@ -0,0 +1,75 @@
+import VanillaInpageNavigation from '@ecl/ec-component-inpage-navigation';
+import demoContent from '@ecl/ec-specs-inpage-navigation/demo/data';
+import loremIpsum from 'lorem-ipsum';
+import Breadcrumb, { BreadcrumbItem } from '@ecl/ec-react-component-breadcrumb';
```
I think that the component example should not include extra components like header and footer; that's the role of the template

planctus commented:
@yhuard , @emeryro , i updated the component to 2.6, adapted all the structure to comply with the new organisation of the repo and the component.

emeryro commented:
Component is working. Still waiting for answers from the design team
